### PR TITLE
Replace prompt templates for AI-Chat and Chat with Group and add jump to PDF viewer

### LINKED ISSUE #167: Add "jump to paper" for AI replies
If AI includes a paper reference, I would like to jump to it.

image

Maybe, the prompt can be changed - or the answer be routed through the AI again :p.

### PR DESCRIPTION
**Pull request by ThiloteE**
### Summary
Replaces the current prompt templates for AI-Chat and Chat with Group.
- Try separate paragraphs for each paper (based on citationkey/snippet pairs) and one for synthesis
- Adds citation links for in-text citations and paragraph appended citations.
- Clicking citation links will open the document viewer
- Differentiate between chatting with multiple papers vs. single papers
- Adds tone and formating

#### Known issues:
Language Models may not perfectly adhere to instructions. As the instructions are very long and detailed, they may forget some of the rules. Maybe we can condense them in a follow up PR.

#### Analogy:
This PR is like honey, because it makes citationkeys stick more often to convenient places. The separation of context resembles the valleys of chocolate bars. Like the sun is shining unto the moon, which is reflected back to earth, users now can get a better glimpse of the underlying citation source, although the view may still be restrained by back-end and rendering shenanigans.

### Steps to test

1. Activate AI features.
2. Configure an AI service provider according to [JabRef's AI-documentation](https://docs.jabref.org/ai/how-to-enable-and-use-ai-features)
3. Link files (PDFs) to multiple entries ([how to](https://docs.jabref.org/collect/findunlinkedfiles#adding-using-drag-and-drop))
4. Add some entries with linked files to a group.
5. Right-Click on Group "Chat with group"
6. Switch the response engine to "Full-Document"
7. Start chatting.
8. Notice that Citation links are appended to paragraphs in responses and there is now a paragraph for each paper in "Chat with group" (tested with 3 papers).

You also can open File > Preferences > AI
Scroll down and have a look at
- System prompt for chatting
- User prompt for chatting

### Related issues and pull requests

Closes NA
Closes https://github.com/InAnYan/jabref/issues/167

### AI usage

AIL3; Mistral Vibe (I believe at present it's either powered by Mistral Medium 3.5 or Mistral small 4)
I created a prompt template and then asked Mistral vibe what parts of the prompt could be turned into an Apache Velocity grammar, then manually adapted the templates further.

Changing the content of jablib\src\main\java\org\jabref\logic\ai\preferences\AiDefaultTemplates.java and opening the pull-request was done manually via Git, Gitui and GitHub.

The other contributors to this PR might have used Ai differently.

<details>
<summary>AI CHECKLIST.md walkthrough</summary>

# Code checklist

> [!IMPORTANT]
> **This is a mandatory final gate.** When the implementation is finished and before you open a PR, work through **every** box below and tick it. If a box cannot be ticked, fix the code first; mark a box `[/]` only if the point genuinely does not apply.
>
> `AGENTS.md` describes how to write code *while* developing — this file confirms the finished result. Do not skip the checklist and do not skip individual points.

## 1. Code self-review

Read your own diff once, top to bottom, and confirm each point.

### Nullability and control flow

- [/] No `== null` / `!= null` checks — JSpecify annotations (`@NullMarked`, `@Nullable`, `@NonNull`) used instead.
- [/] No `Objects.requireNonNull(...)` — nullability expressed via JSpecify annotations.
- [/] New classes annotated with `@NullMarked` (`org.jspecify.annotations.NullMarked`).
- [/] `Optional` consumed with `ifPresent` / `ifPresentOrElse` / `map` / `orElseThrow` — never `orElse(unusedValue)` nor an `isPresent()` + `get()` block.
- [/] `StringUtil.isBlank(...)` used instead of `s == null || s.isBlank()`.

### Exceptions

- [/] No `catch (Exception e)` — only specific exceptions are caught.
- [/] No `throw new RuntimeException(...)` / `IllegalStateException(...)` — these tear down the whole application.
- [/] Logged exceptions are passed as the **last** logger argument (`LOGGER.info("...", e)`), not concatenated into the message string.

### Style and idioms

- [/] New `BibEntry` objects built with withers (`withField`, not `setField`).
- [/] Modern Java used: `List.of()` / `Map.of()` / `Set.of()`, `Path.of()`, `SequencedCollection` / `SequencedSet`, text blocks.
- [/] Regexes use a precompiled `Pattern.compile(...)` constant, not `String.matches(...)`.
- [/] Background work uses `org.jabref.logic.util.BackgroundTask`, not `new Thread()`.
- [/] No commented-out code, no trivial comments restating the code, no AI-disclosure comments in source.
- [/] Markdown Javadoc (`///`) uses Markdown syntax, not JavaDoc inline tags: `` `code` `` instead of `{@code}`, `[ClassName]` instead of `{@link}`.

### User-facing text

- [/] All user-facing text localized (`Localization.lang` in Java, `%` prefix in FXML).
- [/] Sentence case (not Title Case); no trailing `!`; labels do not end with `:`.
- [/] Variance expressed with placeholders (`"...: %0"`), not string concatenation.

### Security

- [/] User-controlled data (request params, entry fields, file contents) is HTML-escaped before being written into any `text/html` response — including exception/error messages, not just the success body (XSS).

### Tests

- [/] Behavior changes in `org.jabref.model` / `org.jabref.logic` have added or updated tests.
- [/] Tests assert object contents (`assertEquals`), use plain JUnit asserts (not AssertJ), have no `@DisplayName`, do not catch exceptions (let them propagate so JUnit reports setup/teardown failures directly), and use `@TempDir` instead of manual temp directories.
- [/] Fetcher tests hit the live endpoints — the remote API is not mocked or stubbed (automated-review suggestions to mock it are rejected on purpose).

## 2. Verification commands

Run in this order — cheapest first. Each must pass.

- [/] `./gradlew :jablib:check` (or `./gradlew check` for all modules).
- [/] `./gradlew checkstyleMain checkstyleTest checkstyleJmh`.
- [/] `./gradlew modernizer`.
- [/] `./gradlew --no-configuration-cache :rewriteDryRun` reports no changes (run `./gradlew rewriteRun` to fix).
- [/] `./gradlew javadoc`.
- [/] `npx markdownlint-cli2 "docs/**/*.md" "*.md"` (only if Markdown changed).
- [/] `npm ci && npm run textlint` reports no misspellings (only if Markdown changed).
- [/] Only if formatting is still off after `rewriteRun`: `docker run -v $(pwd):/github/workspace ghcr.io/leventebajczi/intellij-format:master "*.java" "" ".idea/codeStyles/Project.xml"`.

## 3. Documentation

- [x] `CHANGELOG.md` entry added if the change is visible to the user (end-user wording, no extra blank lines, sorted in next to existing entries about the same component/feature). Link the issue if one exists; link the PR only when no issue exists. Use `TODO` as the placeholder when neither is known yet — never a fake number. No entry for fixes to changes that were themselves introduced after the last release (feature only in `## [Unreleased]`) — update the existing unreleased entry instead if needed.
- [x] Searched [jabref/issues](https://github.com/JabRef/jabref/issues) and [jabref-koppor/issues](https://github.com/JabRef/jabref-koppor/issues) for a related issue; linked only on a confident match, otherwise kept `TODO` (no `closes`/`fixes` for merely-similar issues).
- [/] Requirement added to `docs/requirements/<area>.md` if the change is a new feature or significant bug fix (skip for refactors, minor fixes, and internal changes).
- [/] Developer documentation under `docs/` updated if behavior or architecture changed.

## 4. Pull request

- [x] PR body built from `.github/PULL_REQUEST_TEMPLATE.md`, every section filled.
- [x] All checklist items kept and marked `[x]`, `[ ]`, or `[/]`.
- [x] "Steps to test" is a numbered list with a cropped screenshot of the result for every visible change; no video (only allowed when another program is involved, e.g. drag and drop or push to an external application).
- [x] All HTML comments removed from the PR body.
- [/] PR created with `gh pr create --body-file <file>` (not `--body`).
- [/] If `CHANGELOG.md` used a `TODO` placeholder (no issue confidently identified yet — an existing issue link always stays), the PR was opened as draft, the placeholder was replaced with the real PR-number link after PR creation, committed and pushed, and only then was the PR marked ready for review. If an issue is identified or created later, the link is switched to the issue.

</details>

#### Early version tests:
Prompts and velocity grammar using a syntax with brackets around the citation-key up to commit https://github.com/JabRef/jabref/pull/16981/commits/34372045ce92d4e491fc8a72e0921c2ae1ca2fdb was tested with multiple AI models all fitting into 12GB VRAM, namely:
granite4.2:8b :+1:
https://ollama.com/LiquidAI/lfm2.5-350m :-1:
https://ollama.com/library/lfm2.5 :+1:
https://ollama.com/maternion/ling-3.0-tiny :+1:

#### Later version tests:
The later switch to a citation link was tested with
inclusionAI/Ling-3.0-tiny-UD-Q6_K_XL
Prism-ml/Ternary-Bonsai-27B-Q2_g64

**BEFORE:**
ChatwithGroup-BEFORE-Response1-inclusionAI_Ling-3 0-tiny-UD-Q6_K_XL gguf
ChatwithGroup-BEFORE-Response2-inclusionAI_Ling-3 0-tiny-UD-Q6_K_XL gguf

**AFTER:**
ChatwithGroup-Response1-inclusionAI_Ling-3 0-tiny-UD-Q6_K_XL
ChatwithGroup-Response2-inclusionAI_Ling-3 0-tiny-UD-Q6_K_XL
ChatwithGroup-Response3-inclusionAI_Ling-3 0-tiny-UD-Q6_K_XL
ChatwithGroup-Response4-Prism-ml_Ternary-Bonsai-27B-Q2_g64 gguf
As one can see, the prompt instructions are not always followed perfectly and it involves a little bit of luck. I estimate this will become better with time, once AI service providers receive data from JabRef users and can train on this syntax. Better and larger models might also be better at this than the tiny models I tested it with.

Clicking a citation link will open the linked file with the JabRef internal document viewer:
AI-Chat and Chat with Group - AFTER - Clicking citation link will open Document viewer

### Checklist

- [x] I own the copyright of the code submitted and I license it under the [MIT license](https://github.com/JabRef/jabref/blob/main/LICENSE)
- [x] If AI tools were used, I disclosed them in the "AI usage" section and reviewed, understood, and take full ownership of all AI-generated code
- [x] I manually tested my changes in running JabRef (always required)
- [/] I added JUnit tests for changes (if applicable)
- [ ] I added screenshots in the PR description (if change is visible to the user)
- [x] I added **one sentence (max 20 words)** to `CHANGELOG.md` describing the change from the user's point of view (if the change is visible to the user)
- [/] I checked the [user documentation](https://docs.jabref.org/) for up to dateness and submitted a pull request to our [user documentation repository](https://github.com/JabRef/user-documentation/tree/main/en)

### CODE REVIEW

File: jablib/src/main/java/org/jabref/logic/ai/preferences/AiDefaultTemplates.java [hunk 1]
```diff
@@ -6,22 +6,90 @@
...
-            Here are the papers you are analyzing:
-            #foreach( $entry in $entries )
-            ${CanonicalBibEntry.getCanonicalRepresentation($entry)}
+            You are a helpful research assistant that analyses papers from various scientific fields.
+            You answer questions and to this end will be supplied with snippets from the study and its bibliographic metadata.
+            Relay metadata only, if queried.
+
+            ## Citing sources
+            Each snippet is tagged with a citationkey. The user prompt lists the valid citationkeys for this request and states whether citation is required.
+
```
[the remark this answers is not in the document]
Comment (ThiloteE) [reply jablib/src/main/java/org/jabref/logic/ai/preferences/AiDefaultTemplates.java]: Resolved with commit https://github.com/JabRef/jabref/pull/16981/commits/34372045ce92d4e491fc8a72e0921c2ae1ca2fdb

File: jablib/src/main/java/org/jabref/logic/ai/preferences/AiDefaultTemplates.java [hunk 2]
```diff
@@ -6,22 +6,90 @@
...
+            Being factual and using a professional tone should be a matter of course.
+            Format responses in Markdown.
             #end""";
 
     public static final String CHATTING_USER_MESSAGE_TEMPLATE = """
+            #set( $keys = [] )
+            #foreach( $excerpt in $excerpts )
+            #set( $k = false )
+            #set( $k = $excerpt.source() )
```
[the remark this answers is not in the document]
Comment (ThiloteE) [reply jablib/src/main/java/org/jabref/logic/ai/preferences/AiDefaultTemplates.java]: There are still some issues with rendering the brackets, which might eventually be fixed in a follow up. The regression tests can be added then.

File: CHANGELOG.md [hunk 3]
```diff
@@ -105,6 +105,7 @@ Note that this project **does not** adhere to [Semantic Versioning](https://semv
 - We improved the reliability of online searches by respecting the request limits of arXiv and other services. [#16300](https://github.com/JabRef/jabref/issues/16300)
 - We improved switching between large libraries so entry previews remain responsive while automatic groups and their counts are refreshed. [#16289](https://github.com/JabRef/jabref/pull/16289)
 - We improved the [MODS](https://www.loc.gov/standards/mods/) importer for mapping entry types. [#16055](https://github.com/JabRef/jabref/issues/16055)
+- We changed the default system and user message for chatting (for "AI-Chat" & "Chat with Groups") to better handle etiquette, metadata, citationkeys and context separation. [#16981](https://github.com/JabRef/jabref/pull/16981)
```
Comment (pluto-han) [inline CHANGELOG.md]: Could you please move this entry to the top of `changed`?

**Review by pluto-han — COMMENTED** [review]
just tiny nitpick

File: jablib/src/main/java/org/jabref/logic/ai/preferences/AiDefaultTemplates.java [hunk 4]
```diff
@@ -64,7 +64,7 @@ Valid citationkeys for this request (use only these):
 
             #if( $multiPaper )
             Citation is REQUIRED: append the citationkey(s) at the end of every paragraph that uses snippet content, as specified in the system instructions.
-            Citation is Optional: If in-text citations are used, they have to be formated, as specified in the system instructions.
+            Citation is Optional: If in-text citations are used, they have to be formated as specified in the system instructions.
```
Comment (Siedlerchr) [inline jablib/src/main/java/org/jabref/logic/ai/preferences/AiDefaultTemplates.java]: formatted

**Comment by InAnYan** [comment]
Thank you very much, the PR needs to be updated with the new syntax for citations, I will work on it

File: CHANGELOG.md (hunk 3 above)
Comment (ThiloteE) [reply CHANGELOG.md]: Resolved by https://github.com/JabRef/jabref/pull/16981/commits/33545eda46f3841183d93abe8c9a55d718bed9f8

File: jablib/src/main/java/org/jabref/logic/ai/preferences/AiDefaultTemplates.java (hunk 4 above)
Comment (ThiloteE) [reply jablib/src/main/java/org/jabref/logic/ai/preferences/AiDefaultTemplates.java]: Resolved by https://github.com/JabRef/jabref/pull/16981/commits/cc22a66660dd508ba411cba7c95252a85bddb387

File: CHANGELOG.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (ThiloteE) [reply CHANGELOG.md]: The 20 words rule should be fixed by https://github.com/JabRef/jabref/pull/16981/commits/33545eda46f3841183d93abe8c9a55d718bed9f8

I think the wording is fine. There have been so many changes that it's hard to summarize in a few words. The prompt can be inspected by the user in the preferences, so nothing is really hidden and they can see exactly what changed there and also when using the feature.

File: jablib/src/main/java/org/jabref/logic/ai/preferences/AiDefaultTemplates.java [hunk 5]
```diff
@@ -6,22 +6,90 @@
...
     public static final String CHATTING_USER_MESSAGE_TEMPLATE = """
+            #set( $keys = [] )
+            #foreach( $excerpt in $excerpts )
+            #set( $k = false )
+            #set( $k = $excerpt.source() )
+            #if( $k && !$keys.contains($k) )
+            #set( $added = $keys.add($k) )
+            #end
+            #end
+            #set( $multiPaper = $keys.size() > 1 )
```
[the remark this answers is not in the document]
Comment (ThiloteE) [reply jablib/src/main/java/org/jabref/logic/ai/preferences/AiDefaultTemplates.java]: How do you know that the supplied paper count differs from the supplied key count?

**Comment by ThiloteE** [comment]
I wonder, if we should rename all "citationkey" to "citation key"?

File: jablib/src/main/resources/ai/templates/chatting_user_message.vm [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+## Source materials
+
+Valid citationkeys for this request (use only these):
+#foreach( $k in $keys )
+- $k
+#end
+
+#if( $multiPaper )
+Citation is REQUIRED: append the citation link(s) at the end of every paragraph that uses snippet content, formatted as [@citationkey](entries/citationkey) — or [@citationkey](entries/citationkey#page=N) if that snippet has a page — as specified in the system instructions.
```
Comment (ThiloteE) [inline jablib/src/main/resources/ai/templates/chatting_user_message.vm:19]: If text is copied and pasted, will the end result be [@citationkey] or [citationkey]?
I think the latter is better, if we want to keep compatibility with latex and libreoffice.

**Comment by ThiloteE** [comment]
~~Your additions caused inclusionAI_Ling-3.0-tiny to start thinking in the response (analyzing imaginary "snippets", even though I used the full-document response engine). Last time I tested, there was no thinking in the response, but maybe I was just lucky. Edit: My local setup was inadequate for this test. I had an artificial 2000 token limit for text generation, so that test doesn't contain the end of the thinking. Also, I could have disabled thinking altogether or chosen a non-thinking model.~~

Also, there is a new edge case, where \@ is only at the start with multiple citations. E.g. `[@mustermann1, mustermann2, mustermann3]`

[Test1_inclusionAI_Ling-3.0-tiny-UD-Q6_K_XL_copy and pasted.txt](https://github.com/user-attachments/files/32011728/Test1_inclusionAI_Ling-3.0-tiny-UD-Q6_K_XL_copy.and.pasted.txt)

**Comment by ThiloteE** [comment]
After I fixed my setup and set the "thinking=off" flag in my inference engine, this is how it responded:
image

As you can see, the citations are not at the end anymore and in-text citations are not formated with brackets.
The synthesis paragraph also doesn't have any bracketed citations.

File: jablib/src/main/resources/ai/templates/chatting_user_message.vm (hunk 6 above)
Comment (ThiloteE) [reply jablib/src/main/resources/ai/templates/chatting_user_message.vm:19]: When testing with https://github.com/JabRef/jabref/pull/16981/commits/336ec065fd34a8c179f9fef154dbea2619f8757b, copying and pasting gave `[@citationkey]` as output, except for multi-citations, such as `[@citationkey1, citationkey2, citationkey3]`

Comment (koppor) [reply jablib/src/main/resources/ai/templates/chatting_user_message.vm:19]: The PanDoc manual should have the answer for the correct citation syntax.

**Comment by InAnYan** [comment]
@ThiloteE sorry, it's because of me. Please take a look at the templates in `jablib\src\main\resources\ai\templates`

File: jablib/src/main/resources/ai/templates/chatting_user_message.vm [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+### Metadata
+#foreach( $entry in $entries )
+$!{CanonicalBibEntry.getCanonicalRepresentation($entry)}
+#end
+
+### Snippets
+#foreach( $excerpt in $excerpts )
+#if( $excerpt.pageNumber() )
+<source citationkey="$!{excerpt.source()}" page="$!{excerpt.pageNumber()}">
```
[the remark this answers is not in the document]
Comment (InAnYan) [reply jablib/src/main/resources/ai/templates/chatting_user_message.vm:30]: Yeah, kinda wrong, but nothing bad will happen. In future we seriously (tha'ts a strong word..) planning to add pages

[the remark this answers is not in the document]
Comment (InAnYan) [reply jablib/src/main/resources/ai/templates/chatting_user_message.vm:30]: In practice, llm has never gave me a paged URL

File: jablib/src/main/java/org/jabref/logic/ai/preferences/AiDefaultTemplates.java [hunk 8]
```diff
@@ -1,80 +1,59 @@
...
+    private static String loadTemplate(String templateName) {
+        String resourcePath = TEMPLATE_BASE_PATH + templateName + ".vm";
+        try (var inputStream = AiDefaultTemplates.class.getResourceAsStream(resourcePath)) {
+            if (inputStream == null) {
+                throw new IllegalStateException("Template resource not found: " + resourcePath);
+            }
+            return new String(inputStream.readAllBytes(), StandardCharsets.UTF_8);
+        } catch (IOException e) {
+            LOGGER.error("Failed to load template: {}", resourcePath, e);
+            throw new IllegalStateException("Failed to load template: " + resourcePath, e);
```
[the remark this answers is not in the document]
Comment (InAnYan) [reply jablib/src/main/java/org/jabref/logic/ai/preferences/AiDefaultTemplates.java:53]: well, yeah, but they MUST be present

File: jablib/src/main/java/org/jabref/logic/ai/preferences/AiDefaultTemplates.java [hunk 9]
```diff
@@ -1,80 +1,59 @@
...
+    /// Loads a template from resources.
+    /// @param templateName the name of the template (without extension), e.g., "chatting_system_message"
+    /// @return the template content as a string
+    /// @throws IllegalStateException if the template cannot be loaded
+    private static String loadTemplate(String templateName) {
+        String resourcePath = TEMPLATE_BASE_PATH + templateName + ".vm";
+        try (var inputStream = AiDefaultTemplates.class.getResourceAsStream(resourcePath)) {
+            if (inputStream == null) {
+                throw new IllegalStateException("Template resource not found: " + resourcePath);
+            }
```
[the remark this answers is not in the document]
Comment (InAnYan) [reply jablib/src/main/java/org/jabref/logic/ai/preferences/AiDefaultTemplates.java]: Oh, @wanling0000 could you take a look at this when you have free time? Sorry, if I broke something

File: jablib/src/main/resources/ai/templates/chatting_system_message.vm [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+You answer questions and to this end will be supplied with snippets from the study and its bibliographic metadata.
+Relay metadata only, if queried.
+
+## Citing sources
+Each snippet is tagged with a citationkey, and may optionally be tagged with a page. The user prompt lists the valid citationkeys for this request and states whether citation is required.
+
+Citations are always formatted as a Markdown link of the form:
+[@citationkey](entries/citationkey)
+If — and only if — the snippet you are drawing on was supplied with a page, append a page fragment to the link target instead:
+[@citationkey](entries/citationkey#page=N)
```
[the remark this answers is not in the document]
Comment (InAnYan) [reply jablib/src/main/resources/ai/templates/chatting_system_message.vm:11]: Implemented in https://github.com/JabRef/jabref/pull/16977

**Comment by ThiloteE** [comment]
I think we need to define some terms, stick to them (don't use different terms loosely in the same context) and be really consistent when we refer to them across the prompts, as responses are not very consistent currently (before at https://github.com/JabRef/jabref/pull/16981/commits/34372045ce92d4e491fc8a72e0921c2ae1ca2fdb it was better, at least in my limited testing). I think those terms should be one word, not multiple words.

Maybe a citationkey resembles a citationlink too closely, the usage of citationkey within a link is not well enough defined or there are just too many things in the prompt, which distracts the model from this key difference. `Citation link` might also be too close to the word `citation` and not a good "standalone".

A "paper" is also not well defined. The only two ways for the model to identify what text refers to which paper (or if a chunk is either a whole paper or only a chunk) is by A) deriving this information from within the text and extrapolating, but that's like leaving it up to chance and B) by extrapolating a connection from the position of the citationkey and the text next to it (establishing citationkey text pairs from position/closeness/[grouped punctuation](https://en.wikipedia.org/wiki/Bracket)).

I propose a change of terms to:

1. Snippets = Chunks of text extracted from the papers
2. CitationKey
3. MarkdownCitationLink
4. FullDocument

Maybe it's possible to turn some more of the prompt structure into (pseudo) Velocity syntax.

**Comment by ThiloteE** [comment]
Ah, I misclicked, causing me to close it T.T

**Comment by InAnYan** [comment]
Merged the changes from #16977 into this PR so that the citation prompt formatting and the corresponding clickable citation jump to entry / PDF viewer functionality live together in this pull request.

**Comment by Maran23** [comment]
formatting must be fixed

**Comment by InAnYan** [comment]
sorry, my intellij seems to be broken, can someone help?

File: jablib/src/main/java/org/jabref/logic/ai/preferences/AiDefaultTemplates.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (wanling0000) [reply jablib/src/main/java/org/jabref/logic/ai/preferences/AiDefaultTemplates.java]: Don't worry, the new template resources need to be registered in JabLib’s reachability metadata. May I push the fix directly to this branch?

File: jablib/src/main/java/org/jabref/logic/ai/ingestion/logic/ingestion/FileIngestor.java [hunk 11]
```diff
@@ -25,30 +29,51 @@ public FileIngestor(
...
+        if (pages.isEmpty() || pages.stream().allMatch(String::isBlank)) {
+            LOGGER.error("Unable to generate embeddings for file \"{}\", because JabRef was unable to extract text from the file", path);
             throw new RuntimeException(Localization.lang("Unable to generate embeddings for file '%0', because JabRef was unable to extract text from the file", path.toString()));
         }
+
+        int pageNumber = 1;
+        for (String pageText : pages) {
+            if (Thread.currentThread().isInterrupted()) {
+                throw new InterruptedException();
+            }
```
Comment (Maran23) [inline jablib/src/main/java/org/jabref/logic/ai/ingestion/logic/ingestion/FileIngestor.java:63]: This check here seems weird, why? Can this interrupt?

File: jablib/src/main/resources/ai/templates/summarization_chunk_system_message.vm [hunk 12]
```diff
@@ -0,0 +1,3 @@
+Please provide an overview of the following text. It is a part of a scientific paper.
```
Comment (Maran23) [inline jablib/src/main/resources/ai/templates/summarization_chunk_system_message.vm:1]: Please is usually a waste of takens on AI

File: jabgui/src/main/java/org/jabref/gui/util/SelectableTextFlow.java [hunk 13]
```diff
@@ -162,10 +166,30 @@ private void onMouseClicked(MouseEvent event) {
...
+            return false;
+        }
+
+        Node currentNode = eventTarget;
+        while (currentNode != null && currentNode != this) {
+            if (currentNode instanceof Hyperlink) {
+                return true;
+            }
+            currentNode = currentNode.getParent();
+        }
```
Comment (Maran23) [inline jabgui/src/main/java/org/jabref/gui/util/SelectableTextFlow.java:188]: Two ideas:
- Either check that were not looping the whole scene tree, only to this text flow and then stop, because then we can't be a Hyperlink

- Check how a Hyperlink looks like internally and just do `isHyperlink || paren.isHyperlink || parent.paren.isHyperlink`

File: jabgui/src/main/java/org/jabref/gui/documentviewer/JumpToEntryPdfAction.java [hunk 14]
```diff
@@ -0,0 +1,178 @@
...
+            dialogService.notify(Localization.lang("No PDF files available"));
+            return;
+        }
+
+        openInDocumentViewer(pdfFileOpt.get(), link.page().orElse(1));
+    }
+
+    private static boolean isPdf(LinkedFile file) {
+        try {
+            return FileUtil.isPDFFile(Path.of(file.getLink()));
```
Comment (Maran23) [inline jabgui/src/main/java/org/jabref/gui/documentviewer/JumpToEntryPdfAction.java]: one is `isPdf`, the other one is `isPDF`, casing, decide for one

**Review by Maran23 — DISMISSED** [review]
some things should be improved in a follow-up, but no blocker.

**Comment by wanling0000** [comment]
I pushed a commit directly, hope that's ok... The new `ai/templates/*.vm` resources need to be registered in JabLib's reachability metadata, otherwise `AiDefaultTemplates` fails to initialize in a native image.

File: jablib/src/main/resources/ai/templates/summarization_chunk_system_message.vm (hunk 12 above)
Comment (ThiloteE) [reply jablib/src/main/resources/ai/templates/summarization_chunk_system_message.vm:1]: You are right. The following study found that polite prompts are better than rude prompts, but there is low significance. Only up to +- 3% difference in certain cases. Also depends on the model and domain.
https://arxiv.org/pdf/2512.12812

File: jablib/src/main/java/org/jabref/logic/ai/ingestion/logic/ingestion/FileIngestor.java (hunk 11 above)
Comment (Siedlerchr) [reply jablib/src/main/java/org/jabref/logic/ai/ingestion/logic/ingestion/FileIngestor.java:63]: Yeah I checked, it's running inside a cancellable task  and we want to be able to cancel/interrupt the pdf parsings and embeddings generation

File: jabgui/src/main/java/org/jabref/gui/documentviewer/JumpToEntryPdfAction.java (hunk 14 above)
Comment (koppor) [reply jabgui/src/main/java/org/jabref/gui/documentviewer/JumpToEntryPdfAction.java]: `isPdf` is Google Style - `isPDF` is some JabRef-mixed-style - I don't find a name outside.

I tend to migrate to Google Style, but I don't want to have that much changes in this PR. --> Follow-up.

Thus, please change to `isPDF` for now ^^

File: jabgui/src/main/java/org/jabref/gui/util/SelectableTextFlow.java (hunk 13 above)
Comment (koppor) [reply jabgui/src/main/java/org/jabref/gui/util/SelectableTextFlow.java:188]: We do this in a follow up - our current 🧠 and 🤖 do not come up with something useful for now.  (Filed https://github.com/JabRef/jabref-issue-melting-pot/issues/1370)

**Comment by InAnYan** [comment]
For follow-up PRs:
https://github.com/JabRef/jabref-issue-melting-pot/issues/1371
https://github.com/JabRef/jabref-issue-melting-pot/issues/1372

**Review by koppor — APPROVED** [review]
Thank you for the efforts!

Although there are much code changes and we could certainly improve the code, the general concept is nice and users should be able to try it.

#MoveFast

If it breaks something really, we can revert. I don't think this will happen, because this feature is not seen in the first 15 seconds.

**Review by Maran23 — APPROVED** [review]

**Comment by ThiloteE** [comment]
I will fiddle with the prompt a little in a follow up PR.
Current state:
image

**Comment by InAnYan** [comment]
Yes, in a follow up please!
